### ISSUE TITLE
IE doesn't remember you (was: [ie8] train-2012.07.20 js exception when signing in)

### ISSUE THREAD

**Issue by jrgm** [mozilla/persona#2183]
STR
1. On stage login.anosrep.org and beta.123done.org /sign_in with a clean profile and a known secondary email address
2. Enter the password and hit next

Expected: dialog dismisses and you are logged in

Actual: This JS exception is thrown. You are logged in but 1) if you reload beta.123done.org you're logged in state is not retained, and 2) clicking logout does nothing.

```
Line: 3
Character: 28936
Code: 0
Error Message: Permission denied
URL: https://static.login.anosrep.org/v/1ea909f058/production/en_US/dialog.js
```

In my local dev with train-2012-07.20 the exception in unminified code is:

```
Line: 13
Character: 5
Code: 0
Error Message: Permission denied
URL: http://10.250.2.124:10002/common/js/lib/winchan.js
```

**Comment by jrgm** [comment]
marking 5 star.

**Comment by seanmonstar** [comment]
i'm not able to reproduce... though i am emulating IE8, while I look for a VM. Would it be possible to click Start Debugging in the script debugger, and when the error happens, copy whats in the Call Stack?

**Comment by jrgm** [comment]
I tried with the script debugger, but this is happening on window unload and the ie8 debugger doesn't seem to allow breaking in there. 

So, instead I just wrapped the body of winchan.js removeListener and threw up an alert for arguments.callee.caller. That says the caller is the anonymous function here:

https://github.com/mozilla/browserid/blob/train-2012.07.20/resources/static/common/js/lib/winchan.js#L237

**Comment by jrgm** [comment]
er, wrapped it in a try catch and ....

**Comment by lloyd** [comment]
this will be closed when I merge 0366b9f into train-2012.07.30

**Comment by lloyd** [comment]
renamed this issue, it's still a blocker.  We had mis-attributed the javascript error as the underlying cause previously.  Steps to reproduce are:
1. load up your favorite version of IE
2. point it at beta.123done.org or dev.123done.org
3. sign in
4. reload

Expected: upon reload you're still logged in
Actual: upon reload you get logged out

**Comment by seanmonstar** [comment]
If you put breakpoints in onlogin, onlogout, and onready, you'll notice that after signing in, onlogin gets called as expected, but then right afterwards, onlogout gets called. Only in IE.

I'll go get the kittens.

**Comment by seanmonstar** [comment]
So thanks to some git bisecting from @lloyd, we know that this was introduced with e430a7196d693bf0511c6dff649a89d9f5069907.

The dialog used to set a test cookie, and since max-age didn't delete it in IE8, the communication_iframe could still see it and go along its merry way. That commit removed max-age and manually expires the cookies, which is the right thing to do. It simply exposed that iframes in IE8/9 cannot set cookies, unless it is trusted more.

IE8 uses [P3P](http://en.wikipedia.org/wiki/P3P) to determine iframe trust level, and a solution can be found at [StackOverflow](http://stackoverflow.com/a/389458/101059).
1. We should probably provide a P3P to fix this issue.
2. We should bubble up this low-level `cookies_disabled` issue into something meaningful for the RP, for any other cases that have a similar syptom.

**Comment by lloyd** [comment]
fixed in dev and train-2012.07.30

**Comment by jrgm** [comment]
With IE8/winxp or IE9/win7, with train-2012.07.30 deploy or a dev deploy and 123done:
- can sign in with a existing secondary email address
- after sign in the logout button works
- reloading 123done retains signed-in state
- restarging IE restores signed-in state when I go back to 123done
- if I open a second browser window on 123done: logout and login in one window are reflected in the other window
- can also do the STR for GH-1637 where the logout button would not work after delete-all-history and restarting IE

(I can't test current dev.123done.org at the moment due to an expired certificate (https://bugzilla.mozilla.org/show_bug.cgi?id=780300))

There are some other serious ie8 issues still outstanding that break primaries, signup, forgot password, add email, etc.:

GH-2239 - [ie8] cannot sign in with eyedee.me with ie8 (at least on an ephemeral instance)
GH-2206 - [ie8] javascript exception signing up a new user via /sign_in dialog (or adding or forgot password flows) (dev & train-2012.07.30)
GH-2190 - catch winchan removeListener error in IE8 (which is fixed in dev but not part of train-2012.07.30 atm)

**Comment by jrgm** [comment]
Okay, GH-2206 and GH-2190 are fixed for train-2012.08.17. GH-2190 is no being dealt with in GH-2077
